paywall/src/unlock.js/Wallet.ts

```diff
@@ -0,0 +1,315 @@
+      // if we have no wallet, and no use accounts, we use the web3 proxy wallet
+      this.setupWeb3ProxyWallet()
+    }
+  }
+
+  async enableCryptoWallet() {
+    if (
+      !this.window.web3 ||
+      !this.window.web3.currentProvider ||
+      !this.window.web3.currentProvider.enable
+    ) {
+      return
+    }
+    await this.window.web3.currentProvider.enable()
+  }
+
+  /**
+   * This is the proxy wallet for a crypto wallet
+   *
+   * It handles the "ready" event posted by the Web3ProxyProvider
+   * in the data iframe, and also the "web3" event for
+   * communicating with the crypto wallet
+   */
+  setupWeb3ProxyWallet() {
+    // when receiving a key purchase request, we either pass it to the
+    // account iframe for credit card purchase if user accounts are
+    // explicitly enabled, or to the crypto wallet
+    this.iframes.checkout.on(PostMessages.PURCHASE_KEY, async request => {
+      this.iframes.data.postMessage(PostMessages.PURCHASE_KEY, request)
+    })
+
+    this.iframes.data.on(PostMessages.READY_WEB3, async () => {
+      // initialize, we do this once the iframe is ready to receive information on the wallet
+      // we need to tell the iframe if the wallet is metamask
+      // TODO: pass the name of the wallet if we know it? (secondary importance right now, so omitting)
+      if (!this.hasWallet) {
+        // the user has no crypto wallet
+        this.hasWeb3 = false
+        this.iframes.data.postMessage(PostMessages.WALLET_INFO, {
+          noWallet: true,
+          notEnabled: false,
+          isMetamask: false,
+        })
+        return
+      }
+      try {
+        // first, enable the wallet if necessary
+        await this.enableCryptoWallet()
+      } catch (e) {
+        // user declined to enable the crypto wallet
+        // they still have a wallet, but we need to re-enable it to use it
+        this.hasWeb3 = true
+        this.iframes.data.postMessage(PostMessages.WALLET_INFO, {
+          noWallet: false,
+          notEnabled: true, // user declined to enable the wallet
+          isMetamask: this.isMetamask, // this is used for some decisions in signing
+        })
+        return
+      }
+      this.hasWeb3 = true
+      this.iframes.data.postMessage(PostMessages.WALLET_INFO, {
+        noWallet: false,
+        notEnabled: false,
+        isMetamask: this.isMetamask, // this is used for some decisions in signing
+      })
+    })
+
+    // to communicate with the crypto wallet,
+    // use sendAsync if available, otherwise we will use send
+    const send =
+      this.window.web3 &&
+      this.window.web3.currentProvider &&
+      (this.window.web3.currentProvider.sendAsync ||
+        this.window.web3.currentProvider.send)
+
+    this.iframes.data.on(PostMessages.WEB3, payload => {
+      // handler for the actual web3 calls
+      if (!this.hasWeb3) {
+        // error - no crypto wallet
+        this.iframes.data.postMessage(PostMessages.WEB3_RESULT, {
```

Comment:
I think I completely lost track of why we have to use RPC logic in there. This seems fragile =/ Do you mind providing context?

Reply:
this is the main window side of `Web3ProxyProvider`. Another way of imagining it is that this replaces `src/unlock.js/web3Proxy.ts`. Unlike `web3Proxy.ts`, which intermingled user accounts and crypto wallet handling, this code separates the handling of proxying to the crpyto wallet from proxying to the user accounts iframe.